File: PA58.java
//Leetcode 110. Balanced Binary Tree
//Time complexity: O(n) // n is no of nodes in the tree
//Space Complexity: O(n) // where n is the height of the tree 
class Solution {
    public boolean isBalanced(TreeNode root) {
        if(root==null){
            return true;
        }
        
        int res=helper(root);
        if(res==-1){
            return false;
        }
        return true;
    }
    public int helper(TreeNode root){
       if(root==null){
           return 0;
       }
        
       int lht= helper(root.left);
       int rht= helper(root.right);
        if(lht-rht>1 || rht-lht >1){
            return -1;
        }
        if(lht==-1 || rht==-1){
            return -1;
        }
        return (Math.max(lht,rht)+1);
        
        
        
    }
}

File: PA_58.py
#LC 110 - Balanced Binary Tree
#Time Complexity - O(n)
#Space Complexity - O(h)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        
        def getHeight(root):
            if not root:
                return -1
            left = getHeight(root.left)
            right = getHeight(root.right)
            if left == 9999 or right == 9999:
                return 9999
            if abs(left - right) > 1:
                return 9999
            return max(left,right)+1
        
        if getHeight(root) == 9999:
            return False
        return True